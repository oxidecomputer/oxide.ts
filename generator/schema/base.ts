/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, you can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * Copyright Oxide Computer Company
 */

import { match, P } from "ts-pattern";
import type { OpenAPIV3 } from "openapi-types";
import type { IO } from "../io";

export type Schema = OpenAPIV3.ReferenceObject | OpenAPIV3.SchemaObject;

export const refToSchemaName = (s: string) =>
  s.replace("#/components/schemas/", "");

export interface SchemaHandlers {
  ref: (schema: OpenAPIV3.ReferenceObject, io: IO) => void;
  enum: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  boolean: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  string: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  date: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  number: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  integer: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  array: (schema: OpenAPIV3.ArraySchemaObject, io: IO) => void;
  object: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  oneOf: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  allOf: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  empty: (io: IO) => void;
  default: (schema: Schema, io: IO) => void;
}

export const makeSchemaGenerator =
  (handlers: SchemaHandlers) => (schema: Schema, io: IO) => {
    match(schema)
      .with({ $ref: P.string }, (s) => handlers.ref(s, io))
      .with({ enum: P.array(P.not(P.nullish)) }, (s) => handlers.enum(s, io))
      .with({ type: "boolean" }, (s) => handlers.boolean(s, io))
      .with({ type: "string", format: "date-time" }, (s) =>
        handlers.date(s, io)
      )
      .with({ type: "string" }, (s) => handlers.string(s, io))
      .with({ type: "number" }, (s) => handlers.number(s, io))
      .with({ type: "integer" }, (s) => handlers.integer(s, io))
      .with({ type: "array" }, (s) => handlers.array(s, io))
      .with({ type: "object" }, (s) => handlers.object(s, io))
      .with({ oneOf: P.not(P.nullish) }, (s) => handlers.oneOf(s, io))
      .with({ allOf: P.not(P.nullish) }, (s) => handlers.allOf(s, io))
      .with({}, () => handlers.empty(io))
      .otherwise((s) => handlers.default(s, io));
  };
