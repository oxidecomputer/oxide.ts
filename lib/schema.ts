import { OpenAPIV3 } from "openapi-types";

export type Schema = OpenAPIV3.ReferenceObject | OpenAPIV3.SchemaObject;
export const refToSchemaName = (s: string) =>
  s.replace("#/components/schemas/", "");
