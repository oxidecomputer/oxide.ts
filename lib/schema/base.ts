import { match, P } from "ts-pattern";
import { OpenAPIV3 } from "openapi-types";
import { IO } from "../io";

export type Schema = OpenAPIV3.ReferenceObject | OpenAPIV3.SchemaObject;

export const refToSchemaName = (s: string) =>
  s.replace("#/components/schemas/", "");

export interface SchemaHandlers {
  ref: (schema: OpenAPIV3.ReferenceObject, io: IO) => void;
  enum: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  boolean: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  string: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  date: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  number: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  integer: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  array: (schema: OpenAPIV3.ArraySchemaObject, io: IO) => void;
  object: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  oneOf: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  allOf: (schema: OpenAPIV3.SchemaObject, io: IO) => void;
  empty: (io: IO) => void;
  default: (schema: Schema, io: IO) => void;
}

export const makeSchemaGenerator =
  (handlers: SchemaHandlers) => (schema: Schema, io: IO) => {
    match(schema)
      .with({ $ref: P.string }, (s) => handlers.ref(s, io))
      .with({ enum: P.array(P.not(P.nullish)) }, (s) => handlers.enum(s, io))
      .with({ type: "boolean" }, (s) => handlers.boolean(s, io))
      .with({ type: "string", format: "date-time" }, (s) =>
        handlers.date(s, io)
      )
      .with({ type: "string" }, (s) => handlers.string(s, io))
      .with({ type: "number" }, (s) => handlers.number(s, io))
      .with({ type: "integer" }, (s) => handlers.integer(s, io))
      .with({ type: "array" }, (s) => handlers.array(s, io))
      .with({ type: "object" }, (s) => handlers.object(s, io))
      .with({ oneOf: P.not(P.nullish) }, (s) => handlers.oneOf(s, io))
      .with({ allOf: P.not(P.nullish) }, (s) => handlers.allOf(s, io))
      .with({}, () => handlers.empty(io))
      .otherwise((s) => handlers.default(s, io));
  };
