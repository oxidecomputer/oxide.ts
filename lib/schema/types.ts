import { snakeToCamel } from "../util";
import { makeSchemaGenerator, refToSchemaName } from "./base";

export const schemaToTypes = makeSchemaGenerator({
  ref(schema, { w0 }) {
    if ("$ref" in schema) {
      w0(refToSchemaName(schema.$ref));
    }
  },
  enum(schema, { w0, w }) {
    if (schema.enum!.length === 1) {
      return w0(JSON.stringify(schema.enum![0]));
    }
    for (const arm of schema.enum!) {
      w(`| ${JSON.stringify(arm)}`);
    }
  },
  boolean(_, { w0 }) {
    w0(`boolean`);
  },
  string(schema, { w0 }) {
    w0(`string`);
    if ("nullable" in schema) {
      w0(` | null`);
    }
  },
  date(_, { w0 }) {
    w0(`Date`);
  },
  number(_, { w0 }) {
    w0(`number`);
  },
  integer(_, { w0 }) {
    w0(`number`);
  },
  array(schema, io) {
    schemaToTypes(schema.items, io);
    io.w0(`[]`);
    if ("nullable" in schema) {
      io.w0(` | null`);
    }
  },
  object(schema, io) {
    const { w0, w } = io;
    w0("{");
    for (const [name, subSchema] of Object.entries(schema.properties || {})) {
      const optional = schema.required?.includes(name) ? "" : "?";
      if ("description" in subSchema) {
        w(`\n/** ${subSchema.description} */`);
      }
      w0(`${JSON.stringify(snakeToCamel(name))}${optional}: `);
      schemaToTypes(subSchema, io);
      w0(",");
    }
    w0("}");
  },
  oneOf(schema, io) {
    if (schema.oneOf!.length === 1) {
      return schemaToTypes(schema.oneOf![0], io);
    }
    for (const s of schema.oneOf!) {
      if ("description" in s) {
        io.w(`/** ${s.description} */`);
      }
      io.w0("|");
      schemaToTypes(s, io);
      io.w("");
    }
  },
  allOf(schema, io) {
    if (schema.allOf!.length === 1) {
      return schemaToTypes(schema.allOf![0], io);
    }
    for (const s of schema.allOf!) {
      io.w(`& ${JSON.stringify(s)}`);
    }
  },
  empty({ w0 }) {
    w0(`Record<string, unknown>`);
  },
  default(schema) {
    throw Error(`UNHANDLED SCHEMA: ${JSON.stringify(schema, null, 2)}`);
  },
});
