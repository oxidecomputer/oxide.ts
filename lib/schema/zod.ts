import { IO } from "../io";
import { makeSchemaGenerator, refToSchemaName } from "./base";
import { OpenAPIV3 } from "openapi-types";
import { snakeToCamel } from "../util";

export const schemaToZod = makeSchemaGenerator({
  ref(schema, { w0 }) {
    if ("$ref" in schema) {
      w0(refToSchemaName(schema.$ref));
    }
  },

  boolean(schema, { w0 }) {
    w0(`SafeBoolean`);
    if ("default" in schema) {
      w0(`.default(${schema.default})`);
    }
  },

  enum(schema, io) {
    if (schema.type === "string") {
      io.w0(`z.enum(${JSON.stringify(schema.enum)})`);
    } else if (schema.type === "integer") {
      schemaToZodInt(schema, io);
    } else {
      throw new Error(`Unsupported enum type ${schema.type}`);
    }
  },

  string(schema, { w0 }) {
    w0(`z.string()`);

    if ("default" in schema) {
      w0(`.default(${JSON.stringify(schema.default)})`);
    }
    if (schema.format === "uuid") {
      w0(".uuid()");
    }
    if ("minLength" in schema) {
      w0(`.min(${schema.minLength})`);
    }
    if ("maxLength" in schema) {
      w0(`.max(${schema.maxLength})`);
    }
    if ("pattern" in schema) {
      w0(`.regex(${new RegExp(schema.pattern!).toString()})`);
    }
  },

  date(_, { w0 }) {
    w0("z.coerce.date()");
  },

  number(_, { w0 }) {
    w0("z.number()");
  },

  integer: schemaToZodInt,

  array(schema, io) {
    const { w0 } = io;
    schemaToZod(schema.items, io);
    w0(".array()");
    if ("default" in schema) {
      w0(`.default(${JSON.stringify(schema.default)})`);
    }
  },

  object(schema, io) {
    const { w0, w } = io;
    w0("z.object({");
    for (const [name, subSchema] of Object.entries(schema.properties || {})) {
      w0(`${JSON.stringify(snakeToCamel(name))}: `);
      schemaToZod(subSchema, io);
      if (!schema.required?.includes(name)) {
        w0(`.optional()`);
      }
      w(",");
    }
    w0("})");
  },

  oneOf(schema, io) {
    if (!schema.oneOf) return;
    const { w } = io;

    if (schema.oneOf.length === 1) {
      schemaToZod(schema.oneOf[0], io);
      return;
    }

    /**
     * When dropshot serializes an enum sometimes it breaks it down into a `oneOf` with
     * single element enums such that each individual enum can contain its own description. For zod
     * this translates into a union of single element enums which is unnecessarily complex. We just
     * flatten it down to a single enum here.
     *
     * @see https://github.com/oxidecomputer/oxide.ts/issues/178 for more details
     */
    if (schema.oneOf.every((s) => s && "enum" in s && s.enum?.length === 1)) {
      const enums = schema.oneOf.map(
        (s) => (s as OpenAPIV3.SchemaObject).enum![0]
      );
      w(`z.enum([${enums.map((e) => JSON.stringify(e)).join(", ")}])`);
      return;
    }

    w("z.union([");
    for (const s of schema.oneOf) {
      schemaToZod(s, io);
      w(",");
    }
    w("])");
  },

  allOf(schema, io) {
    if (!schema.allOf) return;
    const { w, w0 } = io;

    if (schema.allOf.length === 0) {
      throw new Error(
        `Unexpected "allOf" should have at least one schema: ${schema}`
      );
    }

    if (schema.allOf.length === 1) {
      schemaToZod(schema.allOf[0], io);
    } else {
      w("z.intersection([");
      for (const s of schema.allOf) {
        schemaToZod(s, io);
        w(",");
      }
      w("])");
    }

    if ("default" in schema) {
      w0(`.default(${JSON.stringify(schema.default)})`);
    }
  },

  empty({ w0 }) {
    w0("z.record(z.unknown())");
  },

  default(schema) {
    throw Error(`UNHANDLED SCHEMA: ${JSON.stringify(schema, null, 2)}`);
  },
});

function schemaToZodInt(schema: OpenAPIV3.SchemaObject, { w0 }: IO) {
  if ("enum" in schema) {
    /**  See comment in {@link setupZod} */
    w0(`IntEnum(${JSON.stringify(schema.enum)} as const)`);
  } else {
    w0(`z.number()`);
  }

  if (schema.default) {
    w0(`.default(${schema.default})`);
  }

  const [, unsigned, size] = schema.format?.match(/(u?)int(\d+)/) || [];
  if ("minimum" in schema) {
    w0(`.min(${schema.minimum})`);
  } else if (unsigned) {
    w0(`.min(0)`);
  } else if (size && parseInt(size) < 64) {
    w0(`.min(-${Math.pow(2, parseInt(size) - 1) - 1})`);
  }

  if ("maximum" in schema) {
    w0(`.max(${schema.maximum})`);
  } else if (size && unsigned && parseInt(size) < 64) {
    w0(`.max(${Math.pow(2, parseInt(size)) - 1})`);
  } else if (size && parseInt(size) < 64) {
    // It's signed so remove the most significant bit
    w0(`.max(${Math.pow(2, parseInt(size) - 1) - 1})`);
  }
}
