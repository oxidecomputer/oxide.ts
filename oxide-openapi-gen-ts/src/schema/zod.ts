/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, you can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * Copyright Oxide Computer Company
 */

import { type IO } from "../io";
import { makeSchemaGenerator, refToSchemaName } from "./base";
import { type OpenAPIV3 } from "openapi-types";
import { snakeToCamel } from "../util";

export const schemaToZod = makeSchemaGenerator({
  ref(schema, { w0 }) {
    if ("$ref" in schema) {
      w0(refToSchemaName(schema.$ref));
    }
  },

  boolean(schema, { w0 }) {
    w0(`SafeBoolean`);
    if ("default" in schema) {
      w0(`.default(${schema.default})`);
    }
    if (schema.nullable) w0(".nullable()");
  },

  enum(schema, io) {
    if (schema.type === "string") {
      io.w0(`z.enum(${JSON.stringify(schema.enum)})`);
    } else if (schema.type === "integer") {
      schemaToZodInt(schema, io);
    } else {
      throw new Error(`Unsupported enum type ${schema.type}`);
    }
    if (schema.nullable) io.w0(".nullable()");
  },

  string(schema, { w0 }) {
    // Handle special formats that become standalone in Zod v4
    if (schema.format === "uuid") {
      w0("z.uuid()");
    } else if (schema.format === "ip") {
      // Generic IP becomes IPv4 for backward compatibility
      w0("z.ipv4()");
    } else if (schema.format === "ipv4") {
      w0("z.ipv4()");
    } else if (schema.format === "ipv6") {
      w0("z.ipv6()");
    } else {
      // Regular string handling
      w0(`z.string()`);

      if ("minLength" in schema) {
        w0(`.min(${schema.minLength})`);
      }
      if ("maxLength" in schema) {
        w0(`.max(${schema.maxLength})`);
      }
      if ("pattern" in schema) {
        w0(`.regex(${new RegExp(schema.pattern!).toString()})`);
      }
    }

    if (schema.nullable) w0(".nullable()");
    if ("default" in schema) w0(`.default(${JSON.stringify(schema.default)})`);
  },

  date(schema, { w0 }) {
    w0("z.coerce.date()");
    if (schema.nullable) w0(".nullable()");
  },

  number(schema, { w0 }) {
    w0("z.number()");
    if (schema.nullable) w0(".nullable()");
  },

  integer(schema, io) {
    schemaToZodInt(schema, io);
    if (schema.nullable) io.w0(".nullable()");
  },

  array(schema, io) {
    const { w0 } = io;
    schemaToZod(schema.items, io);
    w0(".array()");
    if ("default" in schema) {
      w0(`.default(${JSON.stringify(schema.default)})`);
    }
    if (schema.uniqueItems) {
      w0(`.refine(...uniqueItems)`);
    }
  },

  object(schema, io) {
    const { w0, w } = io;
    // record type, which only tells us the type of the values
    if (!schema.properties || Object.keys(schema.properties).length === 0) {
      w0("z.record(z.string(),");
      if (typeof schema.additionalProperties === "object") {
        schemaToZod(schema.additionalProperties, io);
      } else {
        w0("z.unknown()");
      }
      w0(")");
      if (schema.nullable) io.w0(".nullable()");
      return;
    }

    w0("z.object({");
    for (const [name, subSchema] of Object.entries(schema.properties || {})) {
      w0(`${JSON.stringify(snakeToCamel(name))}: `);
      schemaToZod(subSchema, io);
      if (!schema.required?.includes(name)) {
        w0(`.optional()`);
      }
      w(",");
    }
    w0("})");
    if (schema.nullable) io.w0(".nullable()");
  },

  oneOf(schema, io) {
    if (!schema.oneOf) return;
    const { w } = io;

    if (schema.oneOf.length === 1) {
      schemaToZod(schema.oneOf[0], io);
      if (schema.nullable) io.w0(".nullable()");
      return;
    }

    /**
     * When dropshot serializes an enum sometimes it breaks it down into a `oneOf` with
     * single element enums such that each individual enum can contain its own description. For zod
     * this translates into a union of single element enums which is unnecessarily complex. We just
     * flatten it down to a single enum here.
     *
     * @see https://github.com/oxidecomputer/oxide.ts/issues/178 for more details
     */
    if (schema.oneOf.every((s) => s && "enum" in s && s.enum?.length === 1)) {
      const enums = schema.oneOf.map(
        (s) => (s as OpenAPIV3.SchemaObject).enum![0]
      );
      w(`z.enum([${enums.map((e) => JSON.stringify(e)).join(", ")}])`);
      if (schema.nullable) io.w0(".nullable()");
      return;
    }

    w("z.union([");
    for (const s of schema.oneOf) {
      schemaToZod(s, io);
      w(",");
    }
    w("])");
    if (schema.nullable) io.w0(".nullable()");
  },

  allOf(schema, io) {
    if (!schema.allOf) return;
    const { w, w0 } = io;

    if (schema.allOf.length === 0) {
      throw new Error(
        `Unexpected "allOf" should have at least one schema: ${schema}`
      );
    }

    if (schema.allOf.length === 1) {
      schemaToZod(schema.allOf[0], io);
    } else {
      w("z.intersection([");
      for (const s of schema.allOf) {
        schemaToZod(s, io);
        w(",");
      }
      w("])");
    }

    if (schema.nullable) w0(".nullable()");
    if ("default" in schema) w0(`.default(${JSON.stringify(schema.default)})`);
  },

  empty({ w0 }) {
    w0("z.record(z.string(), z.unknown())");
  },

  default(schema) {
    throw Error(`UNHANDLED SCHEMA: ${JSON.stringify(schema, null, 2)}`);
  },
});

function schemaToZodInt(schema: OpenAPIV3.SchemaObject, { w0 }: IO) {
  if ("enum" in schema) {
    /**  See comment in {@link setupZod} */
    w0(`IntEnum(${JSON.stringify(schema.enum)} as const)`);
  } else {
    w0(`z.number()`);
  }

  if (schema.default) {
    w0(`.default(${schema.default})`);
  }

  const [, unsigned, size] = schema.format?.match(/(u?)int(\d+)/) || [];
  if ("minimum" in schema) {
    w0(`.min(${schema.minimum})`);
  } else if (unsigned) {
    w0(`.min(0)`);
  } else if (size && parseInt(size) < 64) {
    w0(`.min(-${Math.pow(2, parseInt(size) - 1) - 1})`);
  }

  if ("maximum" in schema) {
    w0(`.max(${schema.maximum})`);
  } else if (size && unsigned && parseInt(size) < 64) {
    w0(`.max(${Math.pow(2, parseInt(size)) - 1})`);
  } else if (size && parseInt(size) < 64) {
    // It's signed so remove the most significant bit
    w0(`.max(${Math.pow(2, parseInt(size) - 1) - 1})`);
  }
}
