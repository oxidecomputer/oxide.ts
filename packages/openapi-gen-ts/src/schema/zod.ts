/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, you can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * Copyright Oxide Computer Company
 */

import { type IO } from "../io";
import { makeSchemaGenerator, refToSchemaName } from "./base";
import { type OpenAPIV3 } from "openapi-types";
import { snakeToCamel } from "../util";

export const schemaToZod = makeSchemaGenerator({
  ref(schema, { w0 }) {
    if ("$ref" in schema) {
      w0(refToSchemaName(schema.$ref));
    }
  },

  boolean(schema, { w0 }) {
    w0(`SafeBoolean`);
    if ("default" in schema) {
      w0(`.default(${schema.default})`);
    }
  },

  enum(schema, io) {
    if (schema.type === "string") {
      io.w0(`z.enum(${JSON.stringify(schema.enum)})`);
    } else if (schema.type === "integer") {
      schemaToZodInt(schema, io);
    } else {
      throw new Error(`Unsupported enum type ${schema.type}`);
    }
  },

  string(schema, { w0 }) {
    w0(`z.string()`);

    if ("default" in schema) {
      w0(`.default(${JSON.stringify(schema.default)})`);
    }
    if (schema.format === "uuid") {
      w0(".uuid()");
    }

    if (schema.format === "ip") {
      w0(".ip()");
    } else if (schema.format === "ipv4") {
      w0(".ip({ version: 'v4' })");
    } else if (schema.format === "ipv6") {
      w0(".ip({ version: 'v6' })");
    }

    if ("minLength" in schema) {
      w0(`.min(${schema.minLength})`);
    }
    if ("maxLength" in schema) {
      w0(`.max(${schema.maxLength})`);
    }
    if ("pattern" in schema) {
      w0(`.regex(${new RegExp(schema.pattern!).toString()})`);
    }
  },

  date(_, { w0 }) {
    w0("z.coerce.date()");
  },

  number(_, { w0 }) {
    w0("z.number()");
  },

  integer: schemaToZodInt,

  array(schema, io) {
    const { w0 } = io;
    schemaToZod(schema.items, io);
    w0(".array()");
    if ("default" in schema) {
      w0(`.default(${JSON.stringify(schema.default)})`);
    }
    if (schema.uniqueItems) {
      w0(`.refine(...uniqueItems)`);
    }
  },

  object(schema, io) {
    const { w0, w } = io;
    // record type, which only tells us the type of the values
    if (!schema.properties || Object.keys(schema.properties).length === 0) {
      w0("z.record(z.string().min(1),");
      if (typeof schema.additionalProperties === "object") {
        schemaToZod(schema.additionalProperties, io);
      } else {
        w0("z.unknown()");
      }
      w0(")");
      return;
    }

    w0("z.object({");
    for (const [name, subSchema] of Object.entries(schema.properties || {})) {
      w0(`${JSON.stringify(snakeToCamel(name))}: `);
      schemaToZod(subSchema, io);
      if (!schema.required?.includes(name)) {
        w0(`.optional()`);
      }
      w(",");
    }
    w0("})");
  },

  oneOf(schema, io) {
    if (!schema.oneOf) return;
    const { w } = io;

    if (schema.oneOf.length === 1) {
      schemaToZod(schema.oneOf[0], io);
      return;
    }

    /**
     * When dropshot serializes an enum sometimes it breaks it down into a `oneOf` with
     * single element enums such that each individual enum can contain its own description. For zod
     * this translates into a union of single element enums which is unnecessarily complex. We just
     * flatten it down to a single enum here.
     *
     * @see https://github.com/oxidecomputer/oxide.ts/issues/178 for more details
     */
    if (schema.oneOf.every((s) => s && "enum" in s && s.enum?.length === 1)) {
      const enums = schema.oneOf.map(
        (s) => (s as OpenAPIV3.SchemaObject).enum![0]
      );
      w(`z.enum([${enums.map((e) => JSON.stringify(e)).join(", ")}])`);
      return;
    }

    w("z.union([");
    for (const s of schema.oneOf) {
      schemaToZod(s, io);
      w(",");
    }
    w("])");
  },

  allOf(schema, io) {
    if (!schema.allOf) return;
    const { w, w0 } = io;

    if (schema.allOf.length === 0) {
      throw new Error(
        `Unexpected "allOf" should have at least one schema: ${schema}`
      );
    }

    if (schema.allOf.length === 1) {
      schemaToZod(schema.allOf[0], io);
    } else {
      w("z.intersection([");
      for (const s of schema.allOf) {
        schemaToZod(s, io);
        w(",");
      }
      w("])");
    }

    if ("default" in schema) {
      w0(`.default(${JSON.stringify(schema.default)})`);
    }
  },

  empty({ w0 }) {
    w0("z.record(z.unknown())");
  },

  default(schema) {
    throw Error(`UNHANDLED SCHEMA: ${JSON.stringify(schema, null, 2)}`);
  },
});

function schemaToZodInt(schema: OpenAPIV3.SchemaObject, { w0 }: IO) {
  if ("enum" in schema) {
    /**  See comment in {@link setupZod} */
    w0(`IntEnum(${JSON.stringify(schema.enum)} as const)`);
  } else {
    w0(`z.number()`);
  }

  if (schema.default) {
    w0(`.default(${schema.default})`);
  }

  const [, unsigned, size] = schema.format?.match(/(u?)int(\d+)/) || [];
  if ("minimum" in schema) {
    w0(`.min(${schema.minimum})`);
  } else if (unsigned) {
    w0(`.min(0)`);
  } else if (size && parseInt(size) < 64) {
    w0(`.min(-${Math.pow(2, parseInt(size) - 1) - 1})`);
  }

  if ("maximum" in schema) {
    w0(`.max(${schema.maximum})`);
  } else if (size && unsigned && parseInt(size) < 64) {
    w0(`.max(${Math.pow(2, parseInt(size)) - 1})`);
  } else if (size && parseInt(size) < 64) {
    // It's signed so remove the most significant bit
    w0(`.max(${Math.pow(2, parseInt(size) - 1) - 1})`);
  }
}
