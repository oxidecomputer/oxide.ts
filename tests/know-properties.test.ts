import { test, describe, expect } from "vitest";
import { getSchemaTypes } from "./helpers/schema";

/**
 * This test is to help us understand what properties each type in our open API spec are being utilized.
 * If a new property is added but we're _not_ handling it in our generation, we may need to make an update.
 *
 * Review the failing snapshot and then check `zodSchema.ts`. If the new property should be handled, make
 * the necessary changes there.
 */
describe("Known properties", async () => {
  const types = await getSchemaTypes();
  const typeNames = Object.keys(types) as Array<keyof typeof types>;

  test.each(typeNames)(
    "Type %s should contain only known properties",
    (typeName) => {
      const properties = types[typeName]
        .flatMap((s) => Object.keys(s))
        .filter((k, index, arr) => arr.indexOf(k) === index)
        .sort((a, b) => a.localeCompare(b));
      expect(properties).toMatchSnapshot();
    }
  );
});
