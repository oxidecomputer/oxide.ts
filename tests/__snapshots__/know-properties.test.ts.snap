// Vitest Snapshot v1

exports[`Known properties > Type allOf should contain only known properties 1`] = `
[
  "allOf",
  "default",
  "description",
  "nullable",
  "title",
]
`;

exports[`Known properties > Type array should contain only known properties 1`] = `
[
  "default",
  "description",
  "items",
  "nullable",
  "type",
]
`;

exports[`Known properties > Type boolean should contain only known properties 1`] = `
[
  "default",
  "description",
  "type",
]
`;

exports[`Known properties > Type integer should contain only known properties 1`] = `
[
  "description",
  "enum",
  "format",
  "minimum",
  "title",
  "type",
]
`;

exports[`Known properties > Type number should contain only known properties 1`] = `
[
  "format",
  "type",
]
`;

exports[`Known properties > Type object should contain only known properties 1`] = `
[
  "additionalProperties",
  "description",
  "properties",
  "required",
  "type",
]
`;

exports[`Known properties > Type oneOf should contain only known properties 1`] = `
[
  "description",
  "oneOf",
]
`;

exports[`Known properties > Type string should contain only known properties 1`] = `
[
  "default",
  "description",
  "enum",
  "example",
  "format",
  "maxLength",
  "minLength",
  "nullable",
  "pattern",
  "title",
  "type",
]
`;
